streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java

```diff
@@ -0,0 +1,287 @@
+
+import org.apache.kafka.common.utils.SystemTime;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Arrays;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+public class StreamsKafkaClient {
+
+    private static final Logger log = LoggerFactory.getLogger(StreamsKafkaClient.class);
+    private Metadata metadata;
+    private Metrics metrics;
+    private Selector selector;
+    private ChannelBuilder channelBuilder;
+    private KafkaClient kafkaClient;
+    private StreamsConfig streamsConfig;
+    private Node brokerNode;
+
+    public static final String CLEANUP_POLICY_PROP = "cleanup.policy";
+    private static final String COMPACT = "compact";
+
+    private int maxIterations = 10;
+
+    public StreamsKafkaClient(StreamsConfig streamsConfig) {
+
+        this.streamsConfig = streamsConfig;
+        Time time = new SystemTime();
+
+        Map<String, String> metricTags = new LinkedHashMap<>();
+        metricTags.put("client-id", StreamsConfig.CLIENT_ID_CONFIG);
+
+        this.metadata = new Metadata(streamsConfig.getLong(streamsConfig.RETRY_BACKOFF_MS_CONFIG), streamsConfig.getLong(streamsConfig.METADATA_MAX_AGE_CONFIG));
+        List<InetSocketAddress> addresses = ClientUtils.parseAndValidateAddresses(streamsConfig.getList(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG));
+        metadata.update(Cluster.bootstrap(addresses), 0);
+
+        MetricConfig metricConfig = new MetricConfig().samples(streamsConfig.getInt(CommonClientConfigs.METRICS_NUM_SAMPLES_CONFIG))
+                .timeWindow(streamsConfig.getLong(CommonClientConfigs.METRICS_SAMPLE_WINDOW_MS_CONFIG), TimeUnit.MILLISECONDS)
+                .tags(metricTags);
+        List<MetricsReporter> reporters = streamsConfig.getConfiguredInstances(ProducerConfig.METRIC_REPORTER_CLASSES_CONFIG,
+                MetricsReporter.class);
+        // TODO: This should come from the KafkaStream
+        reporters.add(new JmxReporter("kafka.streams"));
+        this.metrics = new Metrics(metricConfig, reporters, time);
+
+        this.channelBuilder = ClientUtils.createChannelBuilder(streamsConfig.values());
+
+        selector = new Selector(streamsConfig.getLong(streamsConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG), this.metrics, time, "kafka-client", this.channelBuilder);
+
+        kafkaClient = new NetworkClient(
+                selector,
+                metadata,
+                streamsConfig.getString(StreamsConfig.CLIENT_ID_CONFIG),
+                100, // a fixed large enough value will suffice
```

Comment:
I used the similar way of creating NetworkClient in KafkaConsumer. For the StreamKafkaClient I can make it a constant as you mentioned. It would be good to update the KafkaConsumer too.